src/core/thread/network_data_publisher.cpp

```diff
@@ -0,0 +1,1016 @@
+}
+
+bool Publisher::DnsSrpServiceEntry::Add(void)
+{
+    // Adds the service entry to the network data.
+
+    bool registerWithLeader = false;
+
+    VerifyOrExit(GetState() == kAdding);
+
+    switch (GetType())
+    {
+    case kTypeAnycast:
+        SuccessOrExit(Get<Service::Manager>().Add<Service::DnsSrpAnycast>(
+            Service::DnsSrpAnycast::ServiceData(mInfo.GetSequenceNumber())));
+        break;
+
+    case kTypeUnicast:
+        SuccessOrExit(Get<Service::Manager>().Add<Service::DnsSrpUnicast>(
+            Service::DnsSrpUnicast::ServiceData(mInfo.GetAddress(), mInfo.GetPort())));
+        break;
+
+    case kTypeUincastMeshLocalEid:
+        SuccessOrExit(Get<Service::Manager>().Add<Service::DnsSrpUnicast>(
+            Service::DnsSrpUnicast::ServerData(mInfo.GetAddress(), mInfo.GetPort())));
+        break;
+    }
+
+    registerWithLeader = true;
+    SetState(kAdded);
+
+exit:
+    return registerWithLeader;
+}
+
+bool Publisher::DnsSrpServiceEntry::Remove(State aNextState)
+{
+    // Removes the service entry from network data (if it was added).
+
+    bool registerWithLeader = false;
+
+    VerifyOrExit((GetState() == kAdded) || (GetState() == kRemoving));
+
+    switch (GetType())
+    {
+    case kTypeAnycast:
+        SuccessOrExit(Get<Service::Manager>().Remove<Service::DnsSrpAnycast>(
+            Service::DnsSrpAnycast::ServiceData(mInfo.GetSequenceNumber())));
+        break;
+
+    case kTypeUnicast:
+        SuccessOrExit(Get<Service::Manager>().Remove<Service::DnsSrpUnicast>(
+            Service::DnsSrpUnicast::ServiceData(mInfo.GetAddress(), mInfo.GetPort())));
+        break;
+
+    case kTypeUincastMeshLocalEid:
+        SuccessOrExit(Get<Service::Manager>().Remove<Service::DnsSrpUnicast>());
+        break;
+    }
+
+    registerWithLeader = true;
+
+    otLogInfoNetData("Publisher: Removed %s from network data", ToString().AsCString());
+
+exit:
+    SetState(aNextState);
+    return registerWithLeader;
+}
+
+void Publisher::DnsSrpServiceEntry::Process(void)
+{
+    // This method checks the entries currently present in Network Data
+    // based on which it then decides whether or not take action
+    // (add/remove or keep monitoring).
+
+    uint8_t numEntries          = 0;
+    uint8_t numPreferredEntries = 0;
+    uint8_t desiredNumEntries   = 0;
+
+    // Do not make any changes if device is not attached, and wait
+    // for role change event.
+    VerifyOrExit(Get<Mle::Mle>().IsAttached());
+
+    VerifyOrExit(GetState() != kNoEntry);
+
+    switch (GetType())
+    {
+    case kTypeAnycast:
+        CountAnycastEntries(numEntries, numPreferredEntries);
+        desiredNumEntries = kDesiredNumAnycast;
+        break;
+
+    case kTypeUnicast:
+    case kTypeUincastMeshLocalEid:
+        CountUnicastEntries(numEntries, numPreferredEntries);
+        desiredNumEntries = kDesiredNumUnicast;
+        break;
+    }
+
+    UpdateState(numEntries, numPreferredEntries, desiredNumEntries);
```

Comment:
What if the service TLV is actually already contained in the `Network Data`, but the entry state is not `kAdded`. The entry may not be able to change to the correct state `kAdded`, and consequently unable to remove it when there are too many service TLVs.

Reply:
I realized that `Publisher` only need to be consistent with Local network data, so no more issue.